File: 2147-number-of-ways-to-divide-a-long-corridor/2147-number-of-ways-to-divide-a-long-corridor.py
class Solution:
    def numberOfWays(self, corridor: str) -> int:
        MOD, count = 1_000_000_007, 1
        
        # Number of seats in current section
        seats = 0
        
        # Tracking index of last S in the previous section
        prev_pair_last = None
        
        # Keep track of seats in the corridor
        for i, thing in enumerate(corridor):
            if thing == "S":
                seats += 1
            
                # If two seats, then this is the last S in the section
                # Update seats for the next section
                if seats == 2:
                    prev_pair_last = i
                    seats = 0
                
                elif seats == 1 and prev_pair_last:
                    count *= i - prev_pair_last
                    count %= MOD
            
        # If odd seats, or zero seats
        if seats == 1 or not prev_pair_last:
            return 0
        
        return count

File: 2147-number-of-ways-to-divide-a-long-corridor/2147-number-of-ways-to-divide-a-long-corridor.cpp
class Solution {
public:
    int numberOfWays(string corridor) {
        const int MOD = 1'000'000'007, n = corridor.size();
        int seats = 0, prevPairLast = -1;
        long count = 1;
        
        for (int i = 0; i < n; ++i){
            if (corridor[i] == 'S'){
                ++seats;
                if (seats == 2){
                    seats = 0;
                    prevPairLast = i;
                }
                else if (seats == 1 && prevPairLast != -1){
                    count = (count * (i - prevPairLast)) % MOD;
                }
            }
        }
        
        // If odd seats, or zero seats
        if (seats == 1 || prevPairLast == -1)
            return 0;
        return (int)count;
    }
};